File: backend/app/schemas/meters.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class SiteOut(BaseModel):
    """Minimal site information returned with a meter."""

    id: int = Field(..., description="Site ID", example=1)
    name: str = Field(..., description="Site name", example="Main Campus")
    slug: Optional[str] = Field(None, description="Site slug", example="main-campus")

    class Config:
        from_attributes = True


class MeterOut(BaseModel):
    """Meter response used by the site/meter filter UI."""

    id: int = Field(..., description="Meter ID", example=1)
    name: str = Field(..., description="Meter name", example="Building A Main Meter")
    serial_number: str = Field(..., description="Meter serial number", example="SN-001")
    model: Optional[str] = Field(None, description="Meter model", example="SmartMeter Pro")
    manufacturer: Optional[str] = Field(None, description="Meter manufacturer", example="VasenVolt")
    meter_type: str = Field(..., description="Meter type", example="electricity")
    status: str = Field(..., description="Meter status", example="active")
    voltage_rating: Optional[float] = Field(None, description="Voltage rating", example=240.0)
    current_rating: Optional[float] = Field(None, description="Current rating", example=100.0)
    power_rating: Optional[float] = Field(None, description="Power rating", example=24000.0)
    accuracy_class: Optional[str] = Field(None, description="Accuracy class", example="0.5")
    installation_date: Optional[datetime] = Field(None, description="Installation date")
    last_calibration: Optional[datetime] = Field(None, description="Last calibration date")
    next_calibration: Optional[datetime] = Field(None, description="Next calibration date")
    communication_protocol: Optional[str] = Field(None, description="Communication protocol", example="MQTT")
    ip_address: Optional[str] = Field(None, description="Meter IP address", example="192.168.1.100")
    port: Optional[int] = Field(None, description="Meter network port", example=1883)
    site_id: int = Field(..., description="Associated site ID", example=1)
    site: SiteOut = Field(..., description="Associated site")
    created_at: datetime = Field(..., description="Meter creation timestamp")
    updated_at: Optional[datetime] = Field(None, description="Meter update timestamp")

    class Config:
        from_attributes = True

File: frontend/src/api/meters.ts
import { apiClient } from './client'

export interface Site {
  id: number
  name: string
  slug?: string
  description?: string | null
  address?: string | null
  city?: string | null
  state?: string | null
  country?: string | null
  postal_code?: string | null
  latitude?: number | null
  longitude?: number | null
  site_type?: string | null
  total_area?: number | null
  is_active?: boolean
}

export interface Meter {
  id: number
  name: string
  serial_number: string
  model?: string | null
  manufacturer?: string | null
  meter_type: string
  status: string
  voltage_rating?: number | null
  current_rating?: number | null
  power_rating?: number | null
  accuracy_class?: string | null
  installation_date?: string | null
  last_calibration?: string | null
  next_calibration?: string | null
  communication_protocol?: string | null
  ip_address?: string | null
  port?: number | null
  site_id: number
  site?: Site | null
  created_at?: string
  updated_at?: string | null
}

export type MetersResponse = Meter[]

export const metersApi = {
  /**
   * Fetch meters list from the /api/meters endpoint
   */
  async getMeters(): Promise<MetersResponse> {
    return apiClient.get<MetersResponse>('/api/meters')
  },
}

